packages/block-editor/src/components/block-inspector/index.js

```diff
@@ -29,19 +39,214 @@ import DefaultStylePicker from '../default-style-picker';
 import useBlockDisplayInformation from '../use-block-display-information';
 import { store as blockEditorStore } from '../../store';
+import BlockIcon from '../block-icon';
+
+function useContentBlocks( blockTypes, block ) {
+	const contenBlocksObjectAux = useMemo( () => {
+		return blockTypes.reduce( ( result, blockType ) => {
+			if (
+				Object.entries( blockType.attributes ).some(
+					( [ , { __experimentalRole } ] ) =>
+						__experimentalRole === 'content'
+				)
+			) {
+				result[ blockType.name ] = true;
+			}
+			return result;
+		}, {} );
+	}, [ blockTypes ] );
+	const isContentBlock = useCallback(
+		( blockName ) => {
+			return !! contenBlocksObjectAux[ blockName ];
+		},
+		[ blockTypes ]
+	);
+	return useMemo( () => {
+		return getContentBlocks( [ block ], isContentBlock );
+	}, [ block, isContentBlock ] );
+}
+
+function getContentBlocks( blocks, isContentBlock ) {
+	const result = [];
+	for ( const block of blocks ) {
+		if ( isContentBlock( block.name ) ) {
+			result.push( block );
+		}
+		result.push( ...getContentBlocks( block.innerBlocks, isContentBlock ) );
+	}
+	return result;
+}
+
+function BlockInspectorNavigationEffects( { children } ) {
+	const { selectBlock } = useDispatch( blockEditorStore );
+	const { goTo, location } = useNavigator();
+	const lastLocationClientId = useRef();
+	const updatingSelectionTo = useRef();
+	const locationClientId = useMemo( () => {
+		if ( location.path.startsWith( '/block/' ) ) {
+			return location.path.substring( '/block/'.length );
+		}
+	}, [ location.path ] );
+	const selectedBlockId = useSelect( ( select ) => {
+		return select( blockEditorStore ).getSelectedBlockClientId();
+	} );
+	// When the location changes update the selection to match the new location.
+	useEffect( () => {
+		lastLocationClientId.current = locationClientId;
+		if ( locationClientId && selectedBlockId !== locationClientId ) {
+			updatingSelectionTo.current = locationClientId;
+			selectBlock( locationClientId );
+		}
+	}, [ locationClientId ] );
+	// When the selection changes update the location to root if no selection update to match location is in progress.
+	useEffect( () => {
+		if ( updatingSelectionTo.current ) {
+			updatingSelectionTo.current = undefined;
+		} else if ( lastLocationClientId.current ) {
+			goTo( '/' );
+		}
+	}, [ selectedBlockId ] );
+
+	return children;
+}
+
+function BlockNavigationButton( { blockTypes, block, selectedBlock } ) {
+	const blockType = blockTypes.find( ( { name } ) => name === block.name );
+	return (
+		<NavigatorButton
+			path={ `/block/${ block.clientId }` }
+			style={
+				selectedBlock.clientId === block.clientId
+					? { color: 'var(--wp-admin-theme-color)' }
+					: {}
+			}
+		>
+			<HStack justify="flex-start">
+				<BlockIcon icon={ blockType.icon } />
+				<FlexItem>{ blockType.title }</FlexItem>
+			</HStack>
+		</NavigatorButton>
+	);
+}
+
+function BlockNavigatorScreen( { block } ) {
```

Comment:
What do you think of extracting these components to its own file? The amount of supporting components/utils is high and introduces noise when one wants to understand the high-level flow.